File: frontend/src/lib/pricing.js
import { DEFAULT_LICENSE_CONFIG, mergeLicenseConfig } from './smsCredits.js';

export const PRICING_CONFIG = { ...DEFAULT_LICENSE_CONFIG };

const PRICING_TIERS = [
  { min: 120000, max: Infinity, startRatio: 0.484375, endRatio: 0.484375 },
  { min: 60000, max: 119999, startRatio: 0.515625, endRatio: 0.484375 },
  { min: 30000, max: 59999, startRatio: 0.539063, endRatio: 0.515625 },
  { min: 15000, max: 29999, startRatio: 0.554688, endRatio: 0.539063 },
  { min: 7500, max: 14999, startRatio: 0.570313, endRatio: 0.554688 },
  { min: 5000, max: 7499, startRatio: 0.601563, endRatio: 0.570313 },
  { min: 2500, max: 4999, startRatio: 0.65625, endRatio: 0.601563 },
  { min: 1500, max: 2499, startRatio: 0.71875, endRatio: 0.65625 },
  { min: 1200, max: 1499, startRatio: 0.804688, endRatio: 0.71875 },
  { min: 750, max: 1199, startRatio: 0.914063, endRatio: 0.804688 },
  { min: 600, max: 749, startRatio: 1.0, endRatio: 0.914063 },
];

const PRODUCT_KEYS = ['communityBuilder', 'engagementBuilder', 'controlTowerUltra'];

export function formatCurrency(amount) {
  if (amount === undefined || amount === null) return '$0';
  const rounded = Math.round(amount);
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 0,
    maximumFractionDigits: 0,
  }).format(rounded);
}

function round2(n) {
  return Math.round(n * 100) / 100;
}

/** Match delphi-me.com calculator: multiply in integer cents. */
function safeMultiply(a, b) {
  const factor = 100;
  return (Math.round(a * factor) * Math.round(b * factor)) / (factor * factor);
}

function getBasePrice(schoolType, licenseConfig) {
  const cfg = licenseConfig || PRICING_CONFIG;
  return cfg[schoolType]?.perStudent ?? cfg.online?.perStudent ?? PRICING_CONFIG.online.perStudent;
}

function getMinimumCost(schoolType, isDistrict, licenseConfig) {
  const cfg = licenseConfig || PRICING_CONFIG;
  const baseMinimum =
    cfg[schoolType]?.minimum ?? cfg.online?.minimum ?? PRICING_CONFIG.online.minimum;
  const districtMinimum = cfg.districtMinimum ?? PRICING_CONFIG.districtMinimum;
  return isDistrict ? Math.max(baseMinimum, districtMinimum) : baseMinimum;
}

function calculateLicenseForProduct(students, schoolType, isDistrict, licenseConfig) {
  const raw = Math.ceil(safeMultiply(students, getBasePrice(schoolType, licenseConfig)));
  return Math.max(raw, getMinimumCost(schoolType, isDistrict, licenseConfig));
}

/** Volume discount for one product — ceil(students × discountPerStudent), matching original. */
function calculateVolumeDiscountPerProduct(students, basePrice) {
  if (students < 500) return 0;
  const tier = PRICING_TIERS.find(t => students >= t.min && students <= t.max);
  if (!tier) return 0;
  let ratio;
  if (tier.max === Infinity) {
    ratio = tier.startRatio;
  } else {
    const progress = (students - tier.min) / (tier.max - tier.min);
    ratio = tier.startRatio + progress * (tier.endRatio - tier.startRatio);
  }
  const pricePerStudent = safeMultiply(basePrice, ratio);
  const discountPerStudent = basePrice - pricePerStudent;
  return Math.ceil(safeMultiply(students, discountPerStudent));
}

function calculateImplementationFee(normalizedSubtotal) {
  if (normalizedSubtotal < 6000) return 1450;
  if (normalizedSubtotal <= 20000) return 1950;
  return 2950;
}

/** Multi-year license discount: yr1 0%, yr2 2.5%, yr3 5%, yr4 7.5%, yr5 10%. */
export function getMultiYearDiscountPercent(years) {
  switch (years) {
    case 2: return 2.5;
    case 3: return 5;
    case 4: return 7.5;
    case 5: return 10;
    default: return 0;
  }
}

function calculateMultiYearDiscount(subtotal, years) {
  const rate = getMultiYearDiscountPercent(years) / 100;
  if (!rate) return 0;
  // Do not use safeMultiply: rates like 2.5%/7.5% round half-up in cents
  // (0.025 → 0.03), inflating the discount (e.g. 7800×2.5% → 234 instead of 195).
  return Math.round(subtotal * rate);
}

function getActiveProducts(quote) {
  return PRODUCT_KEYS.filter(key => Boolean(quote[key]));
}

function computeCustomItemValue(item, percentBase) {
  const amount = Number(item.amount) || 0;
  let value;
  if (item.isPercent) {
    value = safeMultiply(percentBase, amount / 100);
  } else {
    value = amount;
  }
  return item.isDiscount ? -Math.abs(value) : Math.abs(value);
}

function processCustomItems(customItems, annualPercentBase, oneTimePercentBase) {
  return (customItems || []).map(item => {
    const isOneTime = Boolean(item.isOneTime);
    const percentBase = isOneTime ? oneTimePercentBase : annualPercentBase;
    return {
      ...item,
      isOneTime,
      computedValue: computeCustomItemValue(item, percentBase),
    };
  });
}

/**
 * @param {object} quote Full quote state from QuoteContext
 * @param {{ licenseConfig?: object }} [options]
 */
export function calculatePricing(quote, options = {}) {
  const students = Number(quote.students) || 0;
  const years = Number(quote.years) || 1;
  const schoolType = quote.schoolType || 'online';
  const isDistrict = Boolean(quote.isDistrict);
  const isFirstYear = quote.isFirstYear !== false;
  const customItems = quote.customItems || [];
  const licenseConfig = options.licenseConfig
    ? mergeLicenseConfig(options.licenseConfig)
    : PRICING_CONFIG;

  if (students < 0) throw new Error('Students cannot be negative');
  if (years < 1 || years > 5) throw new Error('Years must be between 1 and 5');

  const activeProducts = getActiveProducts(quote);
  const productCount = activeProducts.length;

  const licensePerProduct = productCount > 0
    ? calculateLicenseForProduct(students, schoolType, isDistrict, licenseConfig)
    : 0;

  const productLicenses = {};
  activeProducts.forEach(name => {
    productLicenses[name] = licensePerProduct;
  });

  const productSubtotal = round2(licensePerProduct * productCount);
  const basePrice = getBasePrice(schoolType, licenseConfig);
  const volumePerProduct =
    productCount > 0 ? calculateVolumeDiscountPerProduct(students, basePrice) : 0;
  const volumeDiscount = volumePerProduct * productCount;
  const volumeDiscountPercent =
    productSubtotal > 0 ? Math.round((volumeDiscount / productSubtotal) * 100) : 0;
  const subtotalAfterVolume = round2(productSubtotal - volumeDiscount);

  // Multi-product and multi-year both apply to post-volume subtotal (original order).
  const multiProductDiscount =
    productCount >= 2 ? Math.round(subtotalAfterVolume * 0.1) : 0;
  const multiYearDiscount = calculateMultiYearDiscount(subtotalAfterVolume, years);
  const subtotalAfterMultiProduct = round2(subtotalAfterVolume - multiProductDiscount);
  const subtotalAfterMultiYear = round2(subtotalAfterVolume - multiProductDiscount - multiYearDiscount);

  const normalizedSubtotal = productCount > 0 ? subtotalAfterVolume / productCount : 0;
  const implementationFee = isFirstYear ? calculateImplementationFee(normalizedSubtotal) : 0;

  const cleverFee = quote.clever ? (Number(quote.cleverFee) || 0) : 0;
  const smsFee = quote.sms ? (Number(quote.smsFee) || 0) : 0;
  const addOnTotal = round2(cleverFee + smsFee);

  const oneTimePercentBase = round2(
    (subtotalAfterMultiYear + addOnTotal) * years + implementationFee,
  );
  const processedCustomItems = processCustomItems(
    customItems,
    subtotalAfterMultiProduct,
    oneTimePercentBase,
  );
  const annualCustomTotal = round2(
    processedCustomItems.filter(i => !i.isOneTime).reduce((sum, i) => sum + i.computedValue, 0),
  );
  const dealCustomTotal = round2(
    processedCustomItems.filter(i => i.isOneTime).reduce((sum, i) => sum + i.computedValue, 0),
  );

  const annualBase = round2(subtotalAfterMultiYear + annualCustomTotal + addOnTotal);
  const annualTotal = annualBase;

  const listTotal = round2(productSubtotal + addOnTotal);
  const annualSavings = round2(listTotal - annualTotal);
  const grandTotal = round2(annualTotal * years + implementationFee + dealCustomTotal);
  const listGrandTotal = round2(listTotal * years + implementationFee);
  const totalSavings = round2(listGrandTotal - grandTotal);

  const modulesNet = round2(annualTotal - addOnTotal);
  const perModuleNet = productCount > 0 ? round2(modulesNet / productCount) : 0;
  const volumePerModule = productCount > 0 ? round2(volumeDiscount / productCount) : 0;
  const listPerModule = licensePerProduct;

  const yearBreakdowns = [];
  for (let y = 1; y <= years; y += 1) {
    yearBreakdowns.push({
      year: y,
      license: annualTotal,
      implementation: y === 1 ? implementationFee : 0,
      total: round2(annualTotal + (y === 1 ? implementationFee + dealCustomTotal : 0)),
    });
  }

  return {
    productLicenses,
    productSubtotal,
    volumeDiscount,
    volumeDiscountPercent,
    subtotalAfterVolume,
    multiProductDiscount,
    subtotalAfterMultiProduct,
    subtotalAfterMultiYear,
    implementationFee,
    customItems: processedCustomItems,
    annualCustomTotal,
    dealCustomTotal,
    annualBase,
    multiYearDiscount,
    annualTotal,
    grandTotal,
    years,
    listTotal,
    totalSavings,
    annualSavings,
    cleverFee,
    smsFee,
    addOnTotal,
    yearBreakdowns,
    ebPrice: quote.engagementBuilder ? perModuleNet : 0,
    cbPrice: quote.communityBuilder ? perModuleNet : 0,
    ctuPrice: quote.controlTowerUltra ? perModuleNet : 0,
    listEbPrice: quote.engagementBuilder ? listPerModule : 0,
    listCbPrice: quote.communityBuilder ? listPerModule : 0,
    listCtuPrice: quote.controlTowerUltra ? listPerModule : 0,
    volumePerModule,
    termTotal: round2(annualTotal * years),
    listTermTotal: round2(listTotal * years),
  };
}

/** Default yearly schedule: license split evenly; implementation fee in year 1 only. */
export function buildDefaultYearlyPayments(results) {
  const years = results.years;
  if (years <= 1) return [];

  const licensePerYear = round2(results.termTotal / years);
  const editable = [];
  for (let i = 0; i < years - 1; i += 1) {
    editable.push(
      i === 0
        ? round2(licensePerYear + results.implementationFee + (results.dealCustomTotal || 0))
        : licensePerYear,
    );
  }
  return editable;
}

/**
 * Resolve display schedule for multi-year quotes when pay-upfront is off.
 * Years 1..N-1 come from quote.yearlyPayments (or defaults); year N is the remainder.
 */
export function resolveYearlyPaymentSchedule(quote, results) {
  const years = results.years;
  if (years <= 1 || quote.payUpfront !== false) return [];

  let editable = quote.yearlyPayments;
  if (!Array.isArray(editable) || editable.length !== years - 1) {
    editable = buildDefaultYearlyPayments(results);
  }

  const sumPrevious = editable.reduce((sum, amt) => round2(sum + (Number(amt) || 0)), 0);
  const finalYear = round2(results.grandTotal - sumPrevious);

  const schedule = editable.map((amount, index) => ({
    year: index + 1,
    amount: round2(Number(amount) || 0),
    editable: true,
    isRemainder: false,
  }));
  schedule.push({
    year: years,
    amount: finalYear,
    editable: false,
    isRemainder: true,
  });
  return schedule;
}

/** @deprecated Use calculatePricing(quote) instead */
export class ProductPricingCalculator {
  constructor({ schoolType, students, isDistrict, isFirstYear, years, products, customItems, clever, cleverFee, sms, smsFee }) {
    this.schoolType = schoolType || 'online';
    this.students = Number(students) || 0;
    this.isDistrict = Boolean(isDistrict);
    this.isFirstYear = Boolean(isFirstYear);
    this.years = Number(years) || 1;
    this.products = products || {};
    this.customItems = customItems || [];
    this.clever = Boolean(clever);
    this.cleverFee = cleverFee;
    this.sms = Boolean(sms);
    this.smsFee = smsFee;
  }

  validateInputs() {
    if (this.students < 0) throw new Error('Students cannot be negative');
    if (this.years < 1 || this.years > 5) throw new Error('Years must be between 1 and 5');
  }

  calculateTotal() {
    this.validateInputs();
    return calculatePricing({
      schoolType: this.schoolType,
      students: this.students,
      isDistrict: this.isDistrict,
      isFirstYear: this.isFirstYear,
      years: this.years,
      customItems: this.customItems,
      clever: this.clever,
      cleverFee: this.cleverFee,
      sms: this.sms,
      smsFee: this.smsFee,
      engagementBuilder: this.products.engagementBuilder,
      communityBuilder: this.products.communityBuilder,
      controlTowerUltra: this.products.controlTowerUltra,
    });
  }
}

File: backend/internal/pricing/pricing.go
package pricing

import (
	"fmt"
	"math"
)

type SchoolType string

const (
	SchoolOnline       SchoolType = "online"
	SchoolTraditional  SchoolType = "traditional"
	DistrictMinimum    float64    = 6000
)

type Config struct {
	PerStudent float64
	Minimum    float64
}

var PricingConfig = map[SchoolType]Config{
	SchoolOnline:      {PerStudent: 6.5, Minimum: 3900},
	SchoolTraditional: {PerStudent: 5.0, Minimum: 3000},
}

type Tier struct {
	Min, Max       int
	Start, EndRatio float64
}

var PricingTiers = []Tier{
	{120000, math.MaxInt32, 0.484375, 0.484375},
	{60000, 119999, 0.515625, 0.484375},
	{30000, 59999, 0.539063, 0.515625},
	{15000, 29999, 0.554688, 0.539063},
	{7500, 14999, 0.570313, 0.554688},
	{5000, 7499, 0.601563, 0.570313},
	{2500, 4999, 0.656250, 0.601563},
	{1500, 2499, 0.718750, 0.656250},
	{1200, 1499, 0.804688, 0.718750},
	{750, 1199, 0.914063, 0.804688},
	{600, 749, 1.0, 0.914063},
}

type Products struct {
	EngagementBuilder  bool `json:"engagementBuilder"`
	CommunityBuilder   bool `json:"communityBuilder"`
	ControlTowerUltra  bool `json:"controlTowerUltra"`
	Clever             bool `json:"clever"`
	SMS                bool `json:"sms"`
}

type CustomItem struct {
	ID         int64   `json:"id"`
	Name       string  `json:"name"`
	IsDiscount bool    `json:"isDiscount"`
	IsPercent  bool    `json:"isPercent"`
	IsOneTime  bool    `json:"isOneTime"`
	Amount     float64 `json:"amount"`
}

type QuoteInput struct {
	SchoolType  SchoolType   `json:"schoolType"`
	Students    int          `json:"students"`
	IsDistrict  bool         `json:"isDistrict"`
	IsFirstYear bool         `json:"isFirstYear"`
	Years       int          `json:"years"`
	Products    Products     `json:"products"`
	CustomItems []CustomItem `json:"customItems"`
	SMSFee      float64      `json:"smsFee"`
	CleverFee   float64      `json:"cleverFee"`
	CleverSchools int        `json:"cleverSchools"` // legacy; ignored when CleverFee is used
}

type ModulePrices struct {
	EngagementBuilder float64 `json:"engagementBuilder"`
	CommunityBuilder  float64 `json:"communityBuilder"`
	ControlTowerUltra float64 `json:"controlTowerUltra"`
	Clever            float64 `json:"clever"`
	SMS               float64 `json:"sms"`
}

type YearBreakdown struct {
	Year   int     `json:"year"`
	Amount float64 `json:"amount"`
}

type Result struct {
	ProductLicenses      map[string]float64 `json:"productLicenses"`
	ProductSubtotal      float64            `json:"productSubtotal"`
	VolumeDiscount       float64            `json:"volumeDiscount"`
	MultiProductDiscount float64            `json:"multiProductDiscount"`
	MultiYearDiscount    float64            `json:"multiYearDiscount"`
	ImplementationFee    float64            `json:"implementationFee"`
	AnnualBase           float64            `json:"annualBase"`
	AnnualTotal          float64            `json:"annualTotal"`
	GrandTotal           float64            `json:"grandTotal"`
	ListTotal            float64            `json:"listTotal"`
	TotalSavings         float64            `json:"totalSavings"`
	AnnualSavings        float64            `json:"annualSavings"`
	ModulePrices         ModulePrices       `json:"modulePrices"`
	AddOnTotal           float64            `json:"addOnTotal"`
	Years                int                `json:"years"`
	YearBreakdown        []YearBreakdown    `json:"yearBreakdown"`
}

func round2(v float64) float64 {
	return math.Round(v*100) / 100
}

// safeMul multiplies in integer cents (matches delphi-me.com calculator).
func safeMul(a, b float64) float64 {
	const factor = 100.0
	return (math.Round(a*factor) * math.Round(b*factor)) / (factor * factor)
}

func (q QuoteInput) Validate() error {
	if q.Students < 0 {
		return fmt.Errorf("students cannot be negative")
	}
	if q.Years < 1 || q.Years > 5 {
		return fmt.Errorf("years must be between 1 and 5")
	}
	return nil
}

func (q QuoteInput) basePrice() float64 {
	cfg, ok := PricingConfig[q.SchoolType]
	if !ok {
		cfg = PricingConfig[SchoolOnline]
	}
	return cfg.PerStudent
}

func (q QuoteInput) minimumCost() float64 {
	cfg, ok := PricingConfig[q.SchoolType]
	if !ok {
		cfg = PricingConfig[SchoolOnline]
	}
	return cfg.Minimum
}

func (q QuoteInput) licensePerProduct() float64 {
	raw := math.Ceil(safeMul(float64(q.Students), q.basePrice()))
	return math.Max(raw, q.minimumCost())
}

// volumeDiscountPerProduct: ceil(students × discountPerStudent), matching original calculator.
func volumeDiscountPerProduct(students int, basePrice float64) float64 {
	if students < 500 {
		return 0
	}
	for _, tier := range PricingTiers {
		if students >= tier.Min && students <= tier.Max {
			var ratio float64
			if tier.Max == math.MaxInt32 {
				ratio = tier.Start
			} else {
				progress := float64(students-tier.Min) / float64(tier.Max-tier.Min)
				ratio = tier.Start + progress*(tier.EndRatio-tier.Start)
			}
			pricePerStudent := safeMul(basePrice, ratio)
			discountPerStudent := basePrice - pricePerStudent
			return math.Ceil(safeMul(float64(students), discountPerStudent))
		}
	}
	return 0
}

func implementationFee(normalizedSubtotal float64) float64 {
	if normalizedSubtotal < 6000 {
		return 1450
	}
	if normalizedSubtotal <= 20000 {
		return 1950
	}
	return 2950
}

func multiYearDiscountRate(years int) float64 {
	switch years {
	case 2:
		return 0.025
	case 3:
		return 0.05
	case 4:
		return 0.075
	case 5:
		return 0.10
	default:
		return 0
	}
}

func activePaidModules(p Products) []string {
	var names []string
	if p.EngagementBuilder {
		names = append(names, "engagementBuilder")
	}
	if p.CommunityBuilder {
		names = append(names, "communityBuilder")
	}
	if p.ControlTowerUltra {
		names = append(names, "controlTowerUltra")
	}
	return names
}

func Calculate(q QuoteInput) (Result, error) {
	if err := q.Validate(); err != nil {
		return Result{}, err
	}

	active := activePaidModules(q.Products)
	count := len(active)
	licenseEach := 0.0
	if count > 0 {
		licenseEach = q.licensePerProduct()
	}

	licenses := map[string]float64{}
	for _, name := range active {
		licenses[name] = licenseEach
	}

	productSubtotal := safeMul(licenseEach, float64(count))
	volPerProduct := 0.0
	if count > 0 {
		volPerProduct = volumeDiscountPerProduct(q.Students, q.basePrice())
	}
	volDisc := volPerProduct * float64(count)
	afterVolume := round2(productSubtotal - volDisc)

	// Multi-product and multi-year both apply to post-volume subtotal (original order).
	multiDisc := 0.0
	if count >= 2 {
		multiDisc = math.Round(afterVolume * 0.10)
	}
	afterMulti := round2(afterVolume - multiDisc)

	normalized := 0.0
	if count > 0 {
		normalized = afterVolume / float64(count)
	}
	implementationBase := normalized
	if q.IsDistrict {
		implementationBase = math.Max(normalized, DistrictMinimum)
	}
	implFee := 0.0
	if q.IsFirstYear {
		implFee = implementationFee(implementationBase)
	}

	customTotal := 0.0
	dealCustomTotal := 0.0
	cleverFee := 0.0
	if q.Products.Clever {
		cleverFee = q.CleverFee
	}
	smsFee := 0.0
	if q.Products.SMS {
		smsFee = q.SMSFee
	}
	addOnTotal := round2(cleverFee + smsFee)

	myRate := multiYearDiscountRate(q.Years)
	// Do not use safeMul: rates like 2.5%/7.5% round half-up in cents
	// (0.025 → 0.03), inflating the discount (e.g. 7800×2.5% → 234 instead of 195).
	myDisc := math.Round(afterVolume * myRate)
	afterMultiYear := round2(afterVolume - multiDisc - myDisc)

	oneTimePercentBase := round2((afterMultiYear+addOnTotal)*float64(q.Years) + implFee)
	for _, item := range q.CustomItems {
		base := afterMulti
		if item.IsOneTime {
			base = oneTimePercentBase
		}
		val := item.Amount
		if item.IsPercent {
			val = safeMul(base, item.Amount/100)
		}
		signed := math.Abs(val)
		if item.IsDiscount {
			signed = -signed
		}
		if item.IsOneTime {
			dealCustomTotal += signed
		} else {
			customTotal += signed
		}
	}
	dealCustomTotal = round2(dealCustomTotal)

	annualBase := round2(afterMultiYear + customTotal + addOnTotal)
	annualTotal := annualBase

	listPerYear := round2(productSubtotal + addOnTotal)
	listTerm := round2(listPerYear*float64(q.Years) + implFee)
	grandTotal := round2(annualTotal*float64(q.Years) + implFee + dealCustomTotal)
	totalSavings := round2(listTerm - grandTotal)
	annualSavings := round2(totalSavings / float64(q.Years))

	perModuleAfter := 0.0
	if count > 0 {
		moduleShare := round2(afterMultiYear / float64(count))
		perModuleAfter = moduleShare
		if perModuleAfter < 0 {
			perModuleAfter = 0
		}
	}

	mp := ModulePrices{
		EngagementBuilder: 0,
		CommunityBuilder:  0,
		ControlTowerUltra: 0,
		Clever:            cleverFee,
		SMS:               smsFee,
	}
	if q.Products.EngagementBuilder {
		mp.EngagementBuilder = perModuleAfter
	}
	if q.Products.CommunityBuilder {
		mp.CommunityBuilder = perModuleAfter
	}
	if q.Products.ControlTowerUltra {
		mp.ControlTowerUltra = perModuleAfter
	}

	years := []YearBreakdown{}
	for y := 1; y <= q.Years; y++ {
		amt := annualTotal
		if y == 1 {
			amt = round2(annualTotal + implFee)
		}
		years = append(years, YearBreakdown{Year: y, Amount: amt})
	}

	return Result{
		ProductLicenses:      licenses,
		ProductSubtotal:      productSubtotal,
		VolumeDiscount:       volDisc,
		MultiProductDiscount: multiDisc,
		MultiYearDiscount:    myDisc,
		ImplementationFee:    implFee,
		AnnualBase:           annualBase,
		AnnualTotal:          annualTotal,
		GrandTotal:           grandTotal,
		ListTotal:            listPerYear,
		TotalSavings:         totalSavings,
		AnnualSavings:        annualSavings,
		ModulePrices:         mp,
		AddOnTotal:           addOnTotal,
		Years:                q.Years,
		YearBreakdown:        years,
	}, nil
}
